File: inventory.cpp
#include "inventory.hpp"
#include <iostream>
#include <algorithm>

Inventory::Inventory()
	: m_warehouseName("")
	, m_products()
{
}

Inventory::Inventory(const std::string& warehouseName)
	: m_warehouseName(warehouseName)
	, m_products()
{
}

void Inventory::AddProduct(const Product& product)
{
	m_products.push_back(product);
}

bool Inventory::RemoveProduct(const std::string& productName)
{
	auto it = std::find_if(m_products.begin(), m_products.end(),
		[&productName](const Product& p) {
			return p.GetName() == productName;
		});

	if (it != m_products.end())
	{
		m_products.erase(it);
		return true;
	}
	return false;
}

Product* Inventory::FindProduct(const std::string& productName)
{
	for (auto& product : m_products)
	{
		if (product.GetName() == productName)
		{
			return &product;
		}
	}
	return nullptr;	// Продукт не найден
}

void Inventory::PrintExpiryDates() const
{
	std::cout << "=== Сроки годности продуктов на складе: " << m_warehouseName << " ===" << std::endl;
	
	bool hasExpiredProducts = false;
	
	for (const auto& product : m_products)
	{
		if (!product.CheckExpiryDate())
		{
			hasExpiredProducts = true;
			std::cout << "ВНИМАНИЕ! Продукт '" << product.GetName() 
					  << "' просрочен!" << std::endl;
		}
		else
		{
			int daysLeft = product.GetDaysUntilExpiry();
			std::cout << "Продукт: " << product.GetName() 
					  << " - дней до истечения: " << daysLeft << std::endl;
		}
	}
	
	if (!hasExpiredProducts && m_products.empty())
	{
		std::cout << "Склад пуст." << std::endl;
	}
	else if (!hasExpiredProducts)
	{
		std::cout << "Все продукты в сроке." << std::endl;
	}
	
	std::cout << "===========================================================" << std::endl;
}

void Inventory::PrintInventoryReport() const
{
	std::cout << "=== Отчет о складе: " << m_warehouseName << " ===" << std::endl;
	std::cout << "Количество продуктов на складе: " << m_products.size() << std::endl;
	std::cout << std::endl;
	
	if (m_products.empty())
	{
		std::cout << "Склад пуст." << std::endl;
	}
	else
	{
		for (size_t i = 0; i < m_products.size(); ++i)
		{
			std::cout << "--- Продукт #" << (i + 1) << " ---" << std::endl;
			m_products[i].PrintInfo();
			std::cout << std::endl;
		}
	}
	
	std::cout << "===================================================" << std::endl;
}

std::string Inventory::GetWarehouseName() const
{
	return m_warehouseName;
}

Inventory::ProductsType Inventory::GetProducts() const
{
	return m_products;
}

int Inventory::GetProductsCount() const
{
	return static_cast<int>(m_products.size());
}

void Inventory::SetWarehouseName(const std::string& warehouseName)
{
	m_warehouseName = warehouseName;
}

File: lab4/Inventory.java
import java.util.ArrayList;

public class Inventory {
    // Тип для списка продуктов
    public static class ProductsType extends ArrayList<Product> {
        // Наследуемся от ArrayList для удобства работы
    }

    // Приватные поля
    private String warehouseName;     // Название склада
    private ProductsType products;    // Список продуктов

    // Конструктор по умолчанию
    public Inventory() {
        this.warehouseName = "";
        this.products = new ProductsType();
    }

    // Конструктор с параметрами
    public Inventory(String warehouseName) {
        this.warehouseName = warehouseName;
        this.products = new ProductsType();
    }

    // Добавить продукт на склад
    public void addProduct(Product product) {
        products.add(product);
    }

    // Удалить продукт со склада по имени
    public boolean removeProduct(String productName) {
        for (int i = 0; i < products.size(); i++) {
            if (products.get(i).getName().equals(productName)) {
                products.remove(i);
                return true;
            }
        }
        return false;
    }

    // Найти продукт по имени (возвращает ссылку на продукт)
    public Product findProduct(String productName) {
        for (Product product : products) {
            if (product.getName().equals(productName)) {
                return product;
            }
        }
        return null; // Продукт не найден
    }

    // Вывести информацию о сроках годности всех продуктов
    public void printExpiryDates() {
        System.out.println("=== Сроки годности продуктов на складе: " + warehouseName + " ===");
        
        boolean hasExpiredProducts = false;
        
        for (Product product : products) {
            if (!product.checkExpiryDate()) {
                hasExpiredProducts = true;
                System.out.println("ВНИМАНИЕ! Продукт '" + product.getName() + "' просрочен!");
            } else {
                int daysLeft = product.getDaysUntilExpiry();
                System.out.println("Продукт: " + product.getName() + " - дней до истечения: " + daysLeft);
            }
        }
        
        if (!hasExpiredProducts && products.isEmpty()) {
            System.out.println("Склад пуст.");
        } else if (!hasExpiredProducts) {
            System.out.println("Все продукты в сроке.");
        }
        
        System.out.println("===========================================================");
    }

    // Вывести отчет о хранимых продуктах
    public void printInventoryReport() {
        System.out.println("=== Отчет о складе: " + warehouseName + " ===");
        System.out.println("Количество продуктов на складе: " + products.size());
        System.out.println();
        
        if (products.isEmpty()) {
            System.out.println("Склад пуст.");
        } else {
            for (int i = 0; i < products.size(); i++) {
                System.out.println("--- Продукт #" + (i + 1) + " ---");
                products.get(i).printInfo();
                System.out.println();
            }
        }
        
        System.out.println("===================================================");
    }

    // Геттеры
    public String getWarehouseName() {
        return warehouseName;
    }

    public ProductsType getProducts() {
        return products;
    }

    public int getProductsCount() {
        return products.size();
    }

    // Сеттеры
    public void setWarehouseName(String warehouseName) {
        this.warehouseName = warehouseName;
    }
}

